enhancements/windows-containers/windows-node-egress-proxy.md

```diff
@@ -0,0 +1,351 @@
+The contents of the resource are both user-defined, as well as adjusted by the cluster network operator (CNO). Some
+platforms require instances to access certain endpoints to retrieve metadata for bootstrapping. CNO has logic to inject
+additional `no-proxy` entries such as `169.254.169.254` and `.${REGION}.compute.internal` into the `Proxy` resource.
+
+OLM is a subscriber to these `Proxy` settings -- it forwards the settings to the CSV of managed operators, so the WMCO
+container will automatically get the required `NO_PROXY`, `HTTP_PROXY`, and `HTTPS_PROXY` environment variables on startup.
+In fact, OLM will update and restart the operator pod with proper environment variables if the `Proxy` resource changes.
+
+In proxy-enabled clusters, WMCO will store the 3 proxy variables and their values in the 
+[`windows-services` ConfigMap](./health-management.md#services-configmap).
+In clusters without a global proxy, these variables will not be present in the services ConfigMap.
+This will be done by adding a `EnvironmentVars` key to the top level of the services ConfigMap spec.
+The new key will be an array of name-value pairs representing environment variables to be set on Windows instances.
+```json
+{
+  "EnvironmentVars": [
+    {
+      "name": "name of the environment variable",
+      "value": "value the variable should be set to on the instance"
+    }
+  ],
+  "Services": "<existing spec>",
+  "Files": "<existing spec>",
+}
+```
+
+WICD, which uses the ConfigMap data to configure Windows services, will now have the added responsibility of ensuring
+the `EnvironmentVars` are correctly set on the node and its managed services. WICD will check if any env vars changed by
+comparing each variable's value on the node to the expected value in the ConfigMap. If there is a discrepancy:
+- WICD will update the env vars through a [Windows syscall](https://pkg.go.dev/golang.org/x/sys/windows#SetEnvironmentVariable)
+  which sets the environment variables system-wide, also known as the `Machine` level in Windows. 
+  There are 3 levels to Windows environment variables: `Machine`, `User`, and `Process`. `Process` inherits from 
+  `User`, which inherits from the top level of `Machine`, and sub-processes inherit from their parent process. 
+  In order to route all external node traffic through the proxy, we must set the proxy variables at the `Machine` level.
+  This will also allow the required services (including kubelet and containerd) to inherit the desired values.
+- WICD will reboot the instance so Windows services actually pick up the updated env var values from the OS registry.
+
+### Configuring Custom Trusted Certificates
+
+1. We will add a new ConfigMap to the WMCO's bundle manifests. This resource will contain a trusted CA injection request
+   label so it will be updated by CNO when the global `Proxy` resource changes.
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  annotations:
+    release.openshift.io/create-only: "true"
```

Comment:
Won't this prevent the updates of this configmap with the changed trusted CA bundle contents coming from CNO?

Reply:
Thank you for pointing this out!

As far as I can tell no, it doesn't prevent updates. This is recommended by the [Operator Proxy Support Dev Workflow](https://docs.google.com/document/d/1otp9v5KkoOgq5vhN7ieBpdFhRIIy6z2yuaS_0rjm6wQ/edit#heading=h.dzcnuz4qv07o). It seems this annotation is used to tell CVO not to kick back changes to this resource (i.e. don't delete the CA data injected by CNO).

However, WMCO is managed by OLM, not CVO, and OLM doesn't provide a similar mechanism to ignore changes to manifests. I will update the design to account for this.

Reply:
True, WMCO is OLM managed while this annotation is CVO based. I agree better to remove it to avoid confusion.